tools/flash-tool/flash-tool.8
.\" $Id: flash-tool.8 181 2005-05-07 06:51:15Z pdh $
.\"
.\" Copyright (C) 2004, 2005  Martin Michlmayr <tbm@cyrius.com>
.\" This manual is freely distributable under the terms of the GPL.
.\" It was originally written for Debian GNU/Linux (but may be used
.\" by others).
.\"

.TH FLASH\-TOOL 8 "May 2005"

.SH NAME
flash\-tool \- read or write Cobalt's flash

.SH SYNOPSIS
.B flash\-tool
[\fI\-w\fR] [\fI\-o offset\fR] [\fI\-f\fR] \fIfile\fR

.SH DESCRIPTION
.PP
.B flash\-tool
can be used to read the current content of Cobalt's flash and write it to a
file, or to reprogram the flash with an image supplied by the user.
.B flash\-tool
can therefore be used to first create a backup of Cobalt's original boot
loader and then reprogram the flash with a tested image of colo,
a replacement boot loader for Cobalt which offers more features than the
original boot loader.

.SH OPTIONS
.TP
\fB\-w\fR
Reprogram the flash with the specified image rather than read the current
flash to the specified image (which is the default).

.TP
\fB\-o offset\fR
Specify a specific location of the image as the start.

.SH EXAMPLES

Take a copy of the original boot loader (or whatever your flash currently
contains):
.RS
.ft CW
flash\-tool cobalt\-original\-boot.bin
.ft R
.RE
.PP

Reprogram your flash with another image (see the warning below):
.RS
.ft CW
flash\-tool \-w rom\-image.bin
.ft R
.RE
.PP

.SH WARNING
.PP

Writing a new image to the flash of your Cobalt machine involves
reprogramming the flash.  Do not interrupt the programming process or you
will probably land up with a unit that doesn't boot!  Furthermore, make
sure that the image you write to your flash actually works.  If the
reprogramming fails, you will have to remove and reprogram the flash chip
with an external programmer.  This might mean desoldering the chip as well
as finding a suitable programmer.  Only use the reprogramming functionality
of
.B flash\-tool
with caution and if you know what you are doing.

.SH SEE ALSO
.PP

A detailed description explaining the procedure to make a backup of your
current flash and reprogram it with colo can be found at
.IR /usr/share/doc/colo/INSTALL .

.SH FILES
.PP
.IR /usr/lib/colo/rom\-image.bin
- flashable ROM image of colo

.SH AUTHOR
.PP
Peter Horton <pdh@colonel\-panic.org>



CHANGES
# $Id: CHANGES 195 2006-08-10 22:03:35Z pdh $

CoLo 1.22 (2006-08-10)
----------------------

*	Fix build failure on GCC 4.1.

*	Tweaked INSTALL.
	[Martin Michlmayr]

*	Fixed 'netcon' to not cause an assert in UDP code.

*	Added support for CF cards in CF/IDE adapters (not tested).

CoLo 1.21 (2006-01-22)
----------------------

*	Merged "screen" and "minicom" keymaps.

*	Added "delete word" to line editor ('<ctrl>w').

*	Added 'relocate' command which relocates the initrd image to follow the
	kernel image. Required to use initrd images with current 2.6.x kernels.

*	'mount' command can now be aborted.

CoLo 1.20 (2006-01-16)
----------------------

*	Fix menu cancel/abort/error returns.

*	Update NV format to include keymap.

*	Added keymap for "screen".

*	Wait for network interface transmit ring to drain when disabling network
	interface. This should improve net console output.

*	Added header file dependency checking to stage1, stage2 and chain.

CoLo 1.19 (2005-04-03)
----------------------

*	Fixed stupid bug in ordering of LCD initialisation. This would cause
	installs of 1.17 or 1.18 which weren't an upgrade from 1.16 to crash
	early in stage2.

*	Added "README.tools" to tools directory.

*	Renamed 'md5rom' to 'copy-rom' and added the ability to save a copy of the
	ROM to a file.

*	Moved PCI timeout / retry counter initialisation to PCI initialisation
	code. We now use a retry count of 255 rather than retrying forever. The
	kernel needs a similiar fix.

*	Improved PCI bus scan.

CoLo 1.18 (2005-03-22)
----------------------

*	Added UDP network console support. This adds a 'netcon' command and
	associated variables, and a new boot menu item "Network Shell".

*	Added "ncon" to tools directory. Very simple net console client.

*	PCI serial support switch moved to NV flag 7.

CoLo 1.17 (2005-02-20)
----------------------

*	DHCP client now broadcasts DHCPREQUEST packets. It also includes the
	"server identifier" option from the server in the request. Fixed bug where
	we weren't appending an "end options" tag to the options. [RFC2131]

*	Dropped "no matching socket" diagnostic message.

*	Simplified display of memory size on boot.

*	Added "Cancel" option to 'paneld' menus.

*	Added horizontal scrolling menus to 'paneld' (Controlled by the '-h'
	option).

*	Added horizontal scrolling menus to shell. Controlled by an NV flag (see
	'nvflags' command in README.shell).

*	Reset menu timeout when a button is pressed.

*	Removed ARP diagnostic messages.

CoLo 1.16 (2004-12-16)
----------------------

*	Force menu choice of console on/off on upgrade from earlier versions.

*	Fix 'flash-tool' and 'flash' command to work with TMS29F040 Flash.

CoLo 1.15 (2004-11-28)
----------------------

*	Added link detection to network driver.

*	Added support for 21041 network controller (2700WG units).

*	Added clock speed detection (stage2 only) to ensure timeouts and delays are
	consistent regardless of CPU speed.

*	Added new command 'sleep'.

CoLo 1.14 (2004-10-26)
----------------------

*	Cleaned up build system and added cross compile support.
	[cross compile suggestion / gcc-3.4 link fix - Sebastien Vajda]
	[compiler flag suggestions - Thiemo Seufer]

*	Added new example boot script 'menu.colo' showing new features.

*	Variable expansion in the shell is now carried out before argument
	splitting.  This means variables can now span arguments e.g.

		var message "'first' 'second'"
		lcd {message}

*	Reworked the LCD menu command. The 'menu' command has been removed and
	replaced with a new command, 'select'. The new command justs sets a
	variable to the index of the selected item which can then be used with
	'goto'.

*	Added 'noop' command, which does nothing. When combined with the 'onfail'
	command it can be used for checking the existence of variables.

*	Added script commands 'goto', 'onfail', 'exit' and 'abort'.

*	Added positional marks in scripts using the '@' character.

*	Scripts can now be nested.

*	The 'console-speed' variable is now only available when the serial port is
	enabled.

*	Fixed up support for booting 64-bit kernels (32-bit support was reorganised
	too). 32-bit kernels must be linked for KSEG0 or KSEG1, 64-bit kernels for
	CKSEG0, CKSEG1 or XKPHYS.

*	Ensure we read from Galileo PCI RETRY/TIMEOUT register before writing to
	it. A comment in the kernel source indicates this might be necessary to
	stop Galileo locking up on early units.

*	Added a '-n' command line option to 'putlcd' which allows the user to write
	to a single line of the LCD without affecting the other line.

*	Added big fat warning to 'flash-tool' (just in case).

CoLo 1.13 (2004-08-30)
----------------------

*	Speeded up raw LCD access in LCD tools library.

*	Added 'on' and 'off' options to the 'serial' command.

*	Added 'colo-perm' and 'md5rom' to build.

*	Added missing include to paneld.c.

CoLo 1.12 (2004-08-20)
----------------------

*	The LCD tools 'putlcd', 'paneld' and 'e2fsck-lcd' are now built around a
	common library. The library has support for using direct access to the LCD
	via /dev/mem (the default) or driver access via the Cobalt LCD driver.

	To make the library use the Cobalt LCD driver you need to export an
	environment variable LCD_DEVICE with the name of the device. If LCD_DEVICE
	is an empty string the library will use the default device name, /dev/lcd.

*	Added a '-d' command line option to 'paneld' which makes it fork itself
	into the background on startup.

*	Added 'menu' boot shell command for implementing LCD menus in scripts.
	This feature makes it possible to choose from a list of multiple
	kernels/configurations to boot (see example in README.shell).

*	Fixed a couple of off by one errors in the 'boot' command.

*	Added 'console disable' to the non-volatile flags. This defers initialising
	the serial port until the boot shell is entered and should make it possible
	to use CoLo on units which don't have a serial port. This flag will be set

tools/flash-tool/flash-tool.c
/*
 * (C) P.Horton 2004,2005,2006
 *
 * $Id: flash-tool.c 186 2006-01-17 23:03:58Z pdh $
 *
 * This code is covered by the GNU General Public License. For details see the file "COPYING".
 */

#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>
#include <termios.h>
#include <errno.h>
#include <unistd.h>
#include <fcntl.h>
#include <time.h>
#include <sys/time.h>
#include <sys/mman.h>

#define VER_MAJOR					1
#define VER_MINOR					4

#define APP_NAME					"flash-tool"

#define FLASH_TOTAL_SIZE		(512 << 10)
#define FLASH_BASE_OFFSET		0x1fc00000

#define UNLOCK1(v)				do{FLASH_P[0x5555]=(v);}while(0)
#define UNLOCK2(v)				do{FLASH_P[0x2aaa]=(v);}while(0)
#define UNLOCK1_VALUE			0xaa
#define UNLOCK2_VALUE			0x55

#define CMND_ERASE_SECTOR		0x30
#define CMND_ERASE_SETUP		0x80
#define CMND_AUTOSELECT			0x90
#define CMND_PROGRAM				0xa0
#define CMND_RESET				0xf0

#define ERASE_TIMEOUT			5

#define DEVICE_AM29F040			0x01a4	/* and TMS29F040 */

#define __STR(x)					#x
#define _STR(x)					__STR(x)

static volatile uint8_t *FLASH_P;

static unsigned flash_id(void)
{
	unsigned id;

	/* should give invalid ID if FLASH_P is cached region */

	FLASH_P[0] = CMND_RESET;
	FLASH_P[1] = CMND_RESET;

	UNLOCK1(UNLOCK1_VALUE);
	UNLOCK2(UNLOCK2_VALUE);
	UNLOCK1(CMND_AUTOSELECT);

	id = FLASH_P[0];
	id = FLASH_P[1] | (id << 8);

	UNLOCK1(CMND_RESET);

	return id;
}

static int flash_locked(unsigned long addr)
{
	int lock;

	UNLOCK1(UNLOCK1_VALUE);
	UNLOCK2(UNLOCK2_VALUE);
	UNLOCK1(CMND_AUTOSELECT);

	lock = FLASH_P[addr];

	UNLOCK1(CMND_RESET);

	return lock;
}

static int flash_erase(unsigned long addr)
{
	unsigned nbad, prev, curr, test;
	time_t mark;

	UNLOCK1(UNLOCK1_VALUE);
	UNLOCK2(UNLOCK2_VALUE);
	UNLOCK1(CMND_ERASE_SETUP);
	UNLOCK1(UNLOCK1_VALUE);
	UNLOCK2(UNLOCK2_VALUE);

	FLASH_P[addr] = CMND_ERASE_SECTOR;

	nbad = 0;
	prev = FLASH_P[addr];
	mark = time(NULL);

	for(;;) {

		curr = FLASH_P[addr];

		test = (curr ^ prev) & (1 << 6);
		if(!test)
			break;

		if(prev & curr & (1 << 5)) {
			if(++nbad == 10)
				break;
		} else
			nbad = 0;

		if(time(NULL) - mark > ERASE_TIMEOUT)
			break;

		prev = curr;
	}

	UNLOCK1(CMND_RESET);

	return !test;
}

static int flash_program_byte(unsigned long addr, unsigned data)
{
	unsigned nbad, prev, curr, test;
	time_t mark;

	UNLOCK1(UNLOCK1_VALUE);
	UNLOCK2(UNLOCK2_VALUE);
	UNLOCK1(CMND_PROGRAM);

	FLASH_P[addr] = data;

	nbad = 0;
	prev = FLASH_P[addr];
	mark = time(NULL);

	for(;;) {

		curr = FLASH_P[addr];

		test = (curr ^ prev) & (1 << 6);
		if(!test)
			break;

		if(prev & curr & (1 << 5)) {
			if(++nbad == 10)
				break;
		} else
			nbad = 0;

		if(time(NULL) - mark > 1)
			break;

		prev = curr;
	}

	UNLOCK1(CMND_RESET);

	return !test && FLASH_P[addr] == data;
}

static int flash_program_block(unsigned long addr, const void *data, size_t size)
{
	unsigned indx;

	for(indx = 0; indx < size; ++indx) {

		if(((uint8_t *) data)[indx] & ~FLASH_P[addr + indx]) {

			putchar('*');
			fflush(stdout);

			if(!flash_erase(addr + indx))
				return addr + indx;

			indx = ((addr + indx) | 0xfff) + 1 - addr;
		}
	}

	for(indx = 0; indx < size; ++indx) {

		if(!(indx & 0x1fff)) {
			putchar('+');
			fflush(stdout);
		}

		if(!flash_program_byte(addr + indx, ((uint8_t *) data)[indx]))
			return addr + indx;
	}

	return -1;
}

static int key_wait(void)
{
	struct termios term, save;
	struct timeval tv;
	char match, key;
	int done;

	gettimeofday(&tv, NULL);
	srand48(tv.tv_sec ^ tv.tv_usec ^ getpid());
	match = (unsigned) mrand48() / (unsigned)((0x100000000ULL + 25) / 26) + 'A';

	fputs("\n****************************************************\n"
			"****                                            ****\n"
			"****  THIS WILL ERASE THE CURRENT FIRMWARE AND  ****\n"
			"****                                            ****\n"
			"****  WILL PROBABLY MAKE YOUR UNIT UNBOOTABLE   ****\n"
			"****                                            ****\n"
			"****************************************************\n\n"
			"Are you sure you want to continue (Y/N) ? ", stdout);
	fflush(stdout);

	tcgetattr(0, &term);
	save = term;

	term.c_lflag &= ~(ICANON | ECHO);
	term.c_cc[VMIN] = 0;
	term.c_cc[VTIME] = 4 * 10;
	tcsetattr(0, TCSANOW, &term);

	tcflush(0, TCIFLUSH);
	do
		done = read(0, &key, 1);
	while(done != 1 && errno == EINTR);

	if(key == 'y' || key == 'Y') {

		printf("Y\n\nPress '%c' to proceed ... ", match);
		fflush(stdout);

		tcflush(0, TCIFLUSH);
		do
			done = read(0, &key, 1);
		while(done != 1 && errno == EINTR);

	} else {

		putchar('N');

		key = '\0';
	}
	
	tcsetattr(0, TCSANOW, &save);

	puts("\n");

	return done == 1 && key == match;
}

static void usage(void)
{
	puts("\nusage: " APP_NAME " [ -w ] [ -o offset ] [ -f ] file\n\n  version " _STR(VER_MAJOR) "." _STR(VER_MINOR) "\n");

	exit(1);
}

int main(int argc, char *argv[])
{
	int opt, burn, fd, done, force, trunc;
	unsigned long offset;
	unsigned id, indx;
	off_t size;
	void *data;
	char *ptr;

	burn = 0;
	force = 0;
	offset = 0;
	opterr = 0;
	trunc = 0;

	for(;;) {

		opt = getopt(argc, argv, "o:wFf");
		if(opt == -1)
			break;

		switch(opt) {

			case 'o':
				offset = strtoul(optarg, &ptr, 0);
				if(ptr == optarg || *ptr) {
					fputs(APP_NAME ": invalid offset value\n", stderr);
					return 1;
				}
				break;

			case 'w':
				burn = 1;
				break;

			case 'F':
				force = 1;
				break;

			case 'f':
				trunc = 1;
				break;

			default:
				usage();
		}
	}

	if(argc - optind != 1)
		usage();

	fd = open("/dev/mem", O_RDWR | O_SYNC);
	if(fd == -1) {
		fprintf(stderr, APP_NAME ": failed to open /dev/mem (%s)\n", strerror(errno));
		return 1;
	}

	FLASH_P = mmap(NULL, FLASH_TOTAL_SIZE, PROT_READ | PROT_WRITE, MAP_SHARED, fd, FLASH_BASE_OFFSET);
	if(FLASH_P == MAP_FAILED) {
		fprintf(stderr, APP_NAME ": failed to map /dev/mem (%s)\n", strerror(errno));
		return 1;
	}

	id = flash_id();
	if(id != DEVICE_AM29F040) {
		fprintf(stderr, APP_NAME ": unexpected Flash device 0x%04x\n", id);
		return 1;
	}

	if(burn) {
		
		fd = open(argv[optind], O_RDONLY);
		if(fd == -1) {
			fprintf(stderr, APP_NAME ": failed to open %s (%s)\n", argv[optind], strerror(errno));
			return 1;
		}

		size = lseek(fd, 0, SEEK_END);
		if(size == (off_t) -1) {
			fprintf(stderr, APP_NAME ": seek failed in %s (%s)\n", argv[optind], strerror(errno));
			return 1;
		}

		if(offset + size > FLASH_TOTAL_SIZE) {
			fputs(APP_NAME ": data won't fit in Flash\n", stderr);
			return 1;
		}

		data = mmap(NULL, size, PROT_READ, MAP_SHARED, fd, 0);
		if(data == MAP_FAILED) {
			fprintf(stderr, APP_NAME ": failed to map %s (%s)\n", argv[optind], strerror(errno));
			return 1;
		}

		if(!force && !key_wait()) {
			puts("aborted");
			return 1;
		}

		done = flash_program_block(offset, data, size);

		putchar('\n');

		if(done >= 0) {

			fprintf(stderr, flash_locked(done) ?
				APP_NAME ": programming failed at %06u, *** LOCKED ***\n" :
				APP_NAME ": programming failed at %06u\n",
				done);

			return 1;
		}

		if(memcmp((void *) FLASH_P + offset, data, size)) {
			fputs(APP_NAME ": VERIFY FAILED\n", stderr);
			return 1;
		}

		puts("programmed and verified successfully");

	} else {

		fd = open(argv[optind], O_CREAT | O_TRUNC | O_WRONLY | (trunc ? 0 : O_EXCL), 0664);
		if(fd == -1) {
			fprintf(stderr, APP_NAME ": failed to create %s (%s)\n", argv[optind], strerror(errno));
			return 1;
		}

		for(indx = offset; indx < FLASH_TOTAL_SIZE;) {

			done = write(fd, (void *) FLASH_P + indx, FLASH_TOTAL_SIZE - indx);

			if(done == -1) {
				if(errno != EINTR) {
					fprintf(stderr, APP_NAME ": failed writing %s (%s)\n", argv[optind], strerror(errno));
					return -1;
				}
			} else
				indx += done;
		}

		close(fd);
	}

	return 0;
}

/* vi:set ts=3 sw=3 cin: */
